File: integrationtest/self-healing-integration-tests/src/main/java/de/dkeiss/taxishare/pages/RegistrationPage.java
package de.dkeiss.taxishare.pages;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;

public class RegistrationPage extends AbstractPage {

    public static final String URL = "register";
    public static final String URL_V2 = "registerV2";

    @FindBy(name = "username")
    private WebElement usernameInput;

    @FindBy(name = "email")
    private WebElement emailInput;

    @FindBy(name = "password")
    private WebElement passwordInput;

    @FindBy(className = "btn-primary")
    private WebElement submitButton;

    @FindBy(className = "alert-success")
    private WebElement alertSuccess;

    public RegistrationPage(WebDriver driver) {
        super(driver);
    }

    public void register(String username, String email, String password) {
        usernameInput.sendKeys(username);
        emailInput.sendKeys(email);
        passwordInput.sendKeys(password);
        submitButton.click();
    }

    public String getRegistrationSuccessMessage() {
        return alertSuccess.getText();
    }

    public String getURL() {
        return URL;
    }

}

File: integrationtest_python/self_healing_integration_tests/features/pages/registration_page.py
from __future__ import annotations

from dataclasses import dataclass

from selenium.webdriver.common.by import By

from .base_page import BasePage
from ..support.self_healing import HealableLocator


@dataclass
class RegistrationPage(BasePage):
    url_path: str = "register"

    username_input: HealableLocator = HealableLocator(By.NAME, "username", "Registration username")
    email_input: HealableLocator = HealableLocator(By.NAME, "email", "Registration email")
    password_input: HealableLocator = HealableLocator(By.NAME, "password", "Registration password")
    submit_button: HealableLocator = HealableLocator(By.CSS_SELECTOR, "button.btn-primary", "Register button")
    success_banner: HealableLocator = HealableLocator(By.CSS_SELECTOR, "div.alert-success", "Success message")

    def register(self, username: str, email: str, password: str) -> None:
        self.find(self.username_input).send_keys(username)
        self.find(self.email_input).send_keys(email)
        self.find(self.password_input).send_keys(password)
        self.find(self.submit_button).click()

    def success_message(self) -> str:
        return self.find(self.success_banner).text

